core/commands/cid.go

```diff
@@ -0,0 +1,336 @@
+			verConv: toCidV1,
+		}
+		emitCids(req, resp, opts)
+	},
+	PostRun: cidFmtCmd.PostRun,
+	Type:    cidFmtCmd.Type,
+}
+
+type cidFormatOpts struct {
+	fmtStr  string
+	newBase mbase.Encoding
+	verConv func(cid *cid.Cid) (*cid.Cid, error)
+}
+
+type argumentIterator struct {
+	args []string
+	body cmds.StdinArguments
+}
+
+func (i *argumentIterator) next() (string, bool) {
+	if len(i.args) > 0 {
+		arg := i.args[0]
+		i.args = i.args[1:]
+		return arg, true
+	}
+	if i.body == nil || !i.body.Scan() {
+		return "", false
+	}
+	return strings.TrimSpace(i.body.Argument()), true
+}
+
+func (i *argumentIterator) err() error {
+	if i.body == nil {
+		return nil
+	}
+	return i.body.Err()
+}
+
+func emitCids(req *cmds.Request, resp cmds.ResponseEmitter, opts cidFormatOpts) {
+	itr := argumentIterator{req.Arguments, req.BodyArgs()}
+	for {
+		cidStr, ok := itr.next()
+		if !ok {
+			break
+		}
+		emit := func(fmtd string, err error) {
+			res := &CidFormatRes{CidStr: cidStr, Formatted: fmtd}
+			if err != nil {
+				res.ErrorMsg = err.Error()
+			}
+			resp.Emit(res)
+		}
+		c, err := cid.Decode(cidStr)
+		if err != nil {
+			emit("", err)
+			continue
+		}
+		base := opts.newBase
+		if base == -1 {
+			base, _ = cid.ExtractEncoding(cidStr)
+		}
+		if opts.verConv != nil {
+			c, err = opts.verConv(c)
+			if err != nil {
+				emit("", err)
+				continue
+			}
+		}
+		str, err := cidutil.Format(opts.fmtStr, base, c)
+		if _, ok := err.(cidutil.FormatStringError); ok {
+			// no point in continuing if there is a problem with the format string
+			resp.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+		emit(str, err)
+	}
+	err := itr.err()
+	if err != nil {
+		resp.SetError(err, cmdkit.ErrNormal)
+	}
+}
+
+func toCidV0(c *cid.Cid) (*cid.Cid, error) {
+	if c.Type() != cid.DagProtobuf {
+		return nil, fmt.Errorf("can't convert non-protobuf nodes to cidv0")
+	}
+	return cid.NewCidV0(c.Hash()), nil
+}
+
+func toCidV1(c *cid.Cid) (*cid.Cid, error) {
+	return cid.NewCidV1(c.Type(), c.Hash()), nil
+}
+
+type CodeAndName struct {
+	Code int
+	Name string
+}
+
+var basesCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline: "List available multibase encodings.",
+	},
+	Options: []cmdkit.Option{
+		cmdkit.BoolOption("prefix", "also include the single leter prefixes in addition to the code"),
+		cmdkit.BoolOption("numeric", "also include numeric codes"),
+	},
+	Run: func(req *cmds.Request, resp cmds.ResponseEmitter, env cmds.Environment) {
+		var res []CodeAndName
+		// use EncodingToStr in case at some point there are multiple names for a given code
+		for code, name := range mbase.EncodingToStr {
+			res = append(res, CodeAndName{int(code), name})
+		}
+		cmds.EmitOnce(resp, res)
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeEncoder(func(req *cmds.Request, w io.Writer, val0 interface{}) error {
+			prefixes, _ := req.Options["prefix"].(bool)
+			numeric, _ := req.Options["numeric"].(bool)
+			val := val0.([]CodeAndName)
```

Comment:
For now, you'll need to check if the type is correct. See: https://github.com/ipfs/go-ipfs-cmds/issues/65. Basically, we end up feeding errors to this function as well.

(search the code for examples on how to do this).